File: apps/api/src/routes/sandbox.ts
import { Hono, type Context } from "hono";
import { and, count, eq } from "drizzle-orm";
import { whatsappSessions, type Db } from "@wapi/db";
import {
  fail,
  ok,
  postApiSandboxInboundBody,
  postApiSandboxSessionsBody,
} from "@wapi/contracts";
import {
  encryptSecret,
  generateApiKey,
  generateWebhookSecret,
  hashToken,
  sessionDetailToWire,
  validationFailure,
} from "@wapi/core";
import { gateway, GatewayUnavailableError, SessionNotConnectedError } from "../gateway-client.ts";

/**
 * Sandbox controls — **wapi extensions**, not part of the cloned interface.
 *
 * A sandbox session is a fake number on a fake WhatsApp: it pairs itself, has a small derived
 * directory, accepts sends and can be made to receive them. It exists because linking a real
 * number is the highest-friction step in this product and the one that carries the ban risk, so
 * nobody should have to do it to find out whether their integration works.
 *
 * None of this exists upstream, which is why it lives behind `/api/sandbox/` rather than as
 * flags on documented routes. See `packages/contracts/src/extensions.ts` for the bar each of
 * these had to clear.
 */

/**
 * A cap, not a rate limit.
 *
 * These are free to create through a public endpoint, so the ceiling is what actually stops
 * abuse — an expiry alone would not prevent someone creating thousands in a minute. Deliberately
 * generous: a developer with several branches in flight should not hit it, and being told "too
 * many" is far better than having a session deleted mid-test.
 */
const MAX_SANDBOX_SESSIONS = 25;

export function sandboxRoutes(db: Db) {
  const app = new Hono();

  /**
   * POST /api/sandbox/sessions — create one. Account-scoped, so this needs a PAT.
   *
   * The phone number is not a parameter. It is derived from the session id by the fake engine,
   * under ITU country code 999 which is unassigned and cannot route — letting a caller choose
   * would eventually put a real person's number on a fake session.
   */
  app.post("/sandbox/sessions", async (c) => {
    const auth = c.get("auth");
    if (auth.kind !== "pat") {
      return c.json(fail("Creating a session requires a Personal Access Token."), 403);
    }

    const parsed = postApiSandboxSessionsBody.safeParse(await c.req.json().catch(() => ({})));
    if (!parsed.success) return c.json(validationFailure(parsed.error), 422);

    const [tally] = await db
      .select({ n: count() })
      .from(whatsappSessions)
      .where(
        and(eq(whatsappSessions.accountId, auth.accountId), eq(whatsappSessions.sandbox, true)),
      );
    if ((tally?.n ?? 0) >= MAX_SANDBOX_SESSIONS) {
      return c.json(
        fail(`You already have ${MAX_SANDBOX_SESSIONS} sandbox sessions. Delete one first.`),
        422,
      );
    }

    const apiKey = generateApiKey();
    const [row] = await db
      .insert(whatsappSessions)
      .values({
        accountId: auth.accountId,
        apiKeyEncrypted: encryptSecret(apiKey),
        apiKeyHash: hashToken(apiKey),
        name: parsed.data.name,
        /**
         * A placeholder until the engine derives the real one at connect, which needs the id
         * this insert is about to assign. It is already in the unassigned range, so even this
         * transient value cannot be mistaken for a real number.
         */
        phoneNumber: "+999",
        sandbox: true,
        status: "disconnected",
        webhookSecret: generateWebhookSecret(),
      })
      .returning();

    // Now the id exists, so the derived number can be stored — it is what the dashboard and
    // every API response will show.
    const phoneNumber = `+999${String(row!.id).padStart(8, "0")}`;
    const [session] = await db
      .update(whatsappSessions)
      .set({ phoneNumber })
      .where(eq(whatsappSessions.id, row!.id))
      .returning();

    return c.json(ok(sessionDetailToWire(session!)), 201);
  });

  /**
   * POST /api/sandbox/inbound — fabricate a message *to* this session.
   *
   * The point of the whole feature. It travels the ordinary pipeline, so the webhook that
   * arrives is signed exactly as a real one and your handler cannot tell the difference.
   *
   * Session-scoped: the key identifies which sandbox receives it.
   */
  app.post("/sandbox/inbound", async (c) => {
    const auth = c.get("auth");
    if (auth.kind !== "session") {
      return c.json(fail("This endpoint requires a session API key."), 403);
    }

    const parsed = postApiSandboxInboundBody.safeParse(await c.req.json().catch(() => ({})));
    if (!parsed.success) return c.json(validationFailure(parsed.error), 422);

    try {
      const { key } = await gateway.sandboxInbound(
        auth.sessionId,
        parsed.data.from,
        parsed.data.text,
      );
      return c.json(ok({ key }));
    } catch (err) {
      return sandboxError(c, err);
    }
  });

  /** POST /api/sandbox/scan — finish pairing now instead of waiting for the timer. */
  app.post("/sandbox/scan", async (c) => {
    const auth = c.get("auth");
    if (auth.kind !== "session") {
      return c.json(fail("This endpoint requires a session API key."), 403);
    }
    try {
      await gateway.sandboxScan(auth.sessionId);
      return c.json(ok({ status: "connected" }));
    } catch (err) {
      return sandboxError(c, err);
    }
  });

  return app;
}

/**
 * The gateway refuses these on a session that is not marked sandbox, and that refusal has to
 * reach the caller as something they can act on rather than a 500.
 *
 * A real session hitting a sandbox control is a mistake worth naming precisely: it means the
 * caller believes a live number is fake, which is the more alarming direction of that confusion.
 */
function sandboxError(c: Context, err: unknown) {
  if (err instanceof SessionNotConnectedError) return c.json(fail(err.message), 409);
  if (err instanceof GatewayUnavailableError) {
    return c.json(fail("The WhatsApp service is temporarily unavailable. Please retry."), 503);
  }
  const message = err instanceof Error ? err.message : String(err);
  if (message.includes("not a sandbox session")) {
    return c.json(
      fail("This is not a sandbox session. Sandbox controls only apply to sandbox sessions."),
      422,
    );
  }
  throw err;
}

File: sdk/go/sandbox.go
package wapi

import (
	"context"
	"encoding/json"
)

// Sandbox is a fake number on a fake WhatsApp.
//
// A wapi extension, not part of the WasenderAPI interface. It exists because linking a real
// number is the highest-friction step in this product and the one that carries a ban risk, so you
// should not have to do it to find out whether your integration works.
//
// A sandbox session goes through the same routes and the same code as a real one: it pairs itself
// after a few seconds, has a small deterministic directory, accepts sends and can be made to
// receive them. Its number lives under ITU country code 999, which is unassigned and cannot route
// anywhere.
//
// Two differences from production, deliberate and worth knowing before tuning anything against
// them: account_protection pacing is ignored, so sends return immediately where production would
// wait five seconds; and decrypt-media returns a fixed PNG rather than real media.
type Sandbox struct{ t *transport }

// CreateSession creates a sandbox session. Requires a Personal Access Token.
//
// The number is not yours to choose — it is derived from the session id, so it cannot collide
// with a real one. The response carries the session's own API key, which is what every other call
// here uses.
func (s *Sandbox) CreateSession(ctx context.Context, name string) (*SessionDetail, error) {
	raw, err := s.t.do(ctx, "POST", "/api/sandbox/sessions", nil, map[string]any{"name": name})
	if err != nil {
		return nil, err
	}
	var out SessionDetail
	return &out, unwrap(raw, &out)
}

// Inbound fabricates a message, as if somebody had written to this number.
//
// The reason the sandbox exists. It travels the ordinary pipeline, so the webhook reaching your
// handler is signed exactly as a real one. Use a session key here, not a PAT.
//
// An empty from defaults to the session's first derived contact, so the common case needs no
// sender and the message is still attributable.
func (s *Sandbox) Inbound(ctx context.Context, text, from string) (json.RawMessage, error) {
	body := map[string]any{"text": text}
	if from != "" {
		body["from"] = from
	}
	raw, err := s.t.do(ctx, "POST", "/api/sandbox/inbound", nil, body)
	if err != nil {
		return nil, err
	}
	var out json.RawMessage
	return out, unwrap(raw, &out)
}

// Scan finishes pairing immediately rather than waiting for the fake QR to resolve itself.
//
// Only useful when deliberately testing the waiting state — a sandbox session connects on its own
// a few seconds after Sessions.Connection.Connect.
func (s *Sandbox) Scan(ctx context.Context) error {
	_, err := s.t.do(ctx, "POST", "/api/sandbox/scan", nil, map[string]any{})
	return err
}
